### PR TITLE
[Bug fix] softcap: small inputs no longer vanish in the 1/beta rescale

### LINKED ISSUE #55710: [Bug] ttnn.softcap returns exactly 0 for |x| < beta * 2^-126: the x/beta rescale underflows before tanh
### Describe the bug

`ttnn.softcap(x, beta)` returns exactly `0` for every `|x| < beta * 2**-126`, discarding the input, although `softcap(x) = x` to well under half a bfloat16 ULP over that whole range. The affected band scales with `beta`, so the two betas the op was added for (Gemma logit softcapping, `beta = 30`; Kimi K3 SiTU, `beta = 25`) are the worst cases in the practical range.

The kernel is
([`ckernel_sfpu_softcap.h#L26-L28`](https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_softcap.h#L26-L28)):

```c
sfpi_inline sfpi::vFloat _sfpu_softcap_(sfpi::vFloat x, sfpi::vFloat beta, sfpi::vFloat inv_beta) {
    return _sfpu_tanh_polynomial_(x * inv_beta) * beta;
}
```

`x * inv_beta` is subnormal, and therefore flushed to zero, whenever `|x| < beta * 2**-126`. `tanh(0)` is `0` and `0 * beta` is `0`, so the whole input is lost — even though the answer, `x`, is an ordinary normal float.

### Repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)
x = torch.tensor([[1e-37, 2e-37, 3.5e-37, -1e-37]], dtype=torch.bfloat16)
t = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
print(ttnn.to_torch(ttnn.softcap(t, 30.0)))          # tensor([[0., 0., 0., -0.]], dtype=torch.bfloat16)
print(30.0 * torch.tanh(x.double() / 30.0))          # tensor([[9.9917e-38, 1.9983e-37, 3.4971e-37, -9.9917e-38]])
```

### Observed vs expected (beta = 30)

| input | got | want | ULP |
|---|---|---|---|
| 1.17549e-38 (2^-126) | 0 | 1.175494e-38 | 128 |
| 1e-37 | 0 | 9.991702e-38 | 520 |
| 2e-37 | 0 | 1.998340e-37 | 648 |
| 3.5e-37 | 0 | 3.497096e-37 | 750 |
| 3.6e-37 | 3.599951e-37 | 3.599951e-37 | 0 |

ULP is the integer distance between bfloat16 bit patterns; the relative error over the band is 100%.

Sweeping all 65,536 bfloat16 bit patterns, counting inputs whose golden is a normal float but whose result is exactly zero:

| beta | zeros where the golden is normal | widest \|x\| lost | beta * 2^-126 |
|---|---|---|---|
| 50 | 1426 | 5.87747e-37 | 5.87747e-37 |
| 30 | 1250 | 3.52648e-37 | 3.52648e-37 |
| 25 | 1170 | 2.93874e-37 | 2.93874e-37 |
| 3 | 386 | 3.52648e-38 | 3.52648e-38 |
| 1 | 2 | 1.17549e-38 | 1.17549e-38 |

Max ULP over the whole domain is 753 at `beta = 30` (841 at `beta = 50`); mean ULP is 8.47 (10.63), entirely from this band.

### Expected behavior

`softcap(x, beta) = beta * tanh(x / beta) = x` for small `|x|`, so those inputs should come back unchanged, as they do for `|x| >= beta * 2^-126`.

### Which architectures

Blackhole (`softcap` has no Wormhole kernel). `situ_glu` shares `_sfpu_softcap_` and inherits the same loss on both of its halves.

### How it was verified

ttsim Blackhole v1.10.6 (`TT_METAL_SIMULATOR=libttsim_bh.so`, slow dispatch), exhaustive bfloat16 bit-pattern sweep, golden `beta * tanh(x / beta)` evaluated in float64. `test_softcap_bfloat16_full_domain` does not catch it: it masks out `|golden| <= 1e-30` and only asserts that the region stays below `4e-30`, which returning zero satisfies.

### PR DESCRIPTION
**Pull request by badnikhil**
### PR Category

Bug fix

### Summary

Closes #55710

`_sfpu_softcap_` evaluates `beta * tanh(x * (1/beta))`. The rescale flushes to zero for `|x| < beta * 2^-126`, and `beta * tanh(0)` is exactly 0, so the input was discarded over a band that widens with `beta` — at `beta = 30` every `|x|` below `3.53e-37` came back as `0`, 1250 of the 65,536 bfloat16 bit patterns, although `softcap(x) = x` there.

`tanh(y) = y` to under half an fp32 ULP for `|y| <= 2^-12`, so the kernel now returns `x` unchanged over that range. That covers the whole underflow band and also drops the two roundings the rescale would otherwise introduce, so the near-linear region gets slightly more accurate rather than less.

All 65,536 bfloat16 bit patterns on ttsim Blackhole, golden `beta * tanh(x / beta)` in float64:

| beta | zeros where golden is normal (before → after) | max ULP (before → after) | mean ULP (before → after) |
|---|---|---|---|
| 50 | 1426 → **0** | 841 → **1** | 10.6276 → **0.0030** |
| 30 | 1250 → **0** | 753 → **1** | 8.4702 → **0.0027** |
| 25 | 1170 → **0** | 713 → **1** | 7.5687 → **0.0030** |
| 3 | 386 → **0** | 321 → **1** | 1.3357 → **0.0031** |
| 1 | 2 → **0** | 129 → **1** | 0.0072 → **0.0033** |
| 0.5 | 2 → **0** | 129 → **1** | 0.0072 → **0.0033** |
| 0.001 | 2 → **0** | 129 → **1** | 0.0069 → **0.0030** |

softcap before/after

### Notes for reviewers

- Blackhole only — `softcap` has no Wormhole kernel.
- `situ_glu` calls `_sfpu_softcap_` for both halves and inherits the fix; `test_binary_composite.py -k situ_glu` still passes (9 passed, 1 skipped).
- The threshold is on `y = x * inv_beta`, not on `x`, so it needs no extra multiply and stays correct for any `beta` including the ones where `beta * 2^-12` itself would overflow or underflow.
- `tanh(y) - y = -y^3/3 + O(y^5)`, so at `|y| = 2^-12` the relative difference is `2^-24 / 3`, below half an fp32 ULP; returning `x` is at least as accurate as the polynomial over the whole selected range, which is why the mean ULP drops.
- Edge cases are unchanged: `±inf` still clamps to `±beta`, NaN still does not propagate (the guard's ordered compare is false for NaN, so NaN keeps the existing path), `-0.0` still comes back as `+0.0`. All as documented in the file header.
- `test_softcap_small_magnitude_all_bitpatterns` sweeps every bfloat16 bit pattern for `beta` in `{1, 3, 25, 30, 50}` and asserts the result is bit-identical to the input for `|x| <= beta * 2^-13`, that no normal golden comes back as zero, and 2 ULP overall. It fails all 5 cases on `main`.
- `test_softcap_bfloat16_full_domain` did not catch this: it masks out `|golden| <= 1e-30` and only asserts that region stays below `4e-30`, which returning zero satisfies. That test is left as is.
- Verified on ttsim Blackhole v1.10.6 (slow dispatch); no silicon.

### CODE REVIEW

**Comment by mateusznowakTT** [comment]
@badnikhil : Is the issue fixed here hit in any of the models and affect its accuracy? It adds 30+% performance overhead, so I'd be cautious pulling it in.

**Comment by badnikhil** [comment]
Honest answer: no, I have no evidence any model hits this, and on that basis I would not pull it in at 30%.

The only in-tree caller is the final logit softcapping in `models/experimental/diffusion_gemma` (`final_logit_softcapping = 30.0`), applied to logits out of the LM head. Those are order 1 to 30. The band that breaks is `|x| < beta * 2^-126`, i.e. below 3.5e-37 at beta = 30, so nothing a real logit distribution produces. I found it by sweeping the full bfloat16 domain, not from a model regression, and I should have said that in the description rather than leaving the impact open.

Two things worth weighing before you close it:

- The guard fires for `|x| <= beta * 2^-12` (about 7.3e-3 at beta = 30), not only for the underflowing band, because that is the whole region where `tanh(y) = y` to under half an fp32 ULP. So it also removes the two roundings the rescale introduces there. That is a real but small accuracy gain in a range models do reach.
- A branch-free version is not available here. The value is destroyed by the underflow inside `x * inv_beta`, and any pre-scaling that avoids it is only valid in the linear region, so recovering it needs a compare either way. If the cost is unacceptable the honest options are to drop it or to put it behind a compile-time flag so the default path is untouched.

Happy to do whichever you prefer: close this, or respin it as a template parameter defaulted off. If you would rather keep the op fast and simply document the limit, say so and I will close it and leave #55710 open with the numbers so the behaviour is at least recorded.

One check I could not run: I do not have silicon, so the 30% is your measurement, not something I can confirm or improve against. If you can point me at the microbenchmark you used I will measure any alternative before proposing it.
